lock.go

```
@@ -0,0 +1,240 @@
+package backupstore
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/longhorn/backupstore/util"
+)
+
+const (
+	LOCKS_DIRECTORY       = "locks"
+	LOCK_PREFIX           = "lock"
+	LOCK_SUFFIX           = ".lck"
+	LOCK_DURATION         = time.Second * 150
+	LOCK_MAX_WAIT_TIME    = time.Second * 30
+	LOCK_REFRESH_INTERVAL = time.Second * 60
+	LOCK_CHECK_INTERVAL   = time.Second * 10
+	LOCK_CHECK_WAIT_TIME  = time.Second * 2
+)
+
+type LockType int
+
+const UNTYPED_LOCK LockType = 0
+const BACKUP_LOCK LockType = 1
+const RESTORE_LOCK LockType = 1
+const DELETION_LOCK LockType = 2
+
+type FileLock struct {
+	Name       string
+	Type       LockType
+	Acquired   bool
+	driver     BackupStoreDriver
+	volume     string
+	count      int32
+	serverTime time.Time
+	keepAlive  chan struct{}
+	mutex      sync.Mutex
+}
+
+func New(driver BackupStoreDriver, volumeName string, lockType LockType) (*FileLock, error) {
+	return &FileLock{driver: driver, volume: volumeName,
+		Type: lockType, Name: util.GenerateName(LOCK_PREFIX)}, nil
+}
+
+// isExpired checks whether the current lock is expired
+func (lock *FileLock) isExpired() bool {
+	isExpired := time.Now().Sub(lock.serverTime) > LOCK_DURATION
+	return isExpired
+}
+
+func (lock *FileLock) canAcquire() bool {
+	canAcquire := true
+	locks := getLocksForVolume(lock.volume, lock.driver)
+	for _, serverLock := range locks {
+		serverLockHasDifferentType := serverLock.Type != lock.Type
+		serverLockHasPriority := compareLocks(serverLock, lock) < 0
+		if serverLockHasDifferentType && serverLockHasPriority && !serverLock.isExpired() {
+			canAcquire = false
+			break
+		}
+	}
+
+	return canAcquire
+}
+
+func (lock *FileLock) Lock() error {
+	lock.mutex.Lock()
+	defer lock.mutex.Unlock()
+	if lock.Acquired {
+		atomic.AddInt32(&lock.count, 1)
+		_ = saveLock(lock)
```

I am not sure if we can directly ignore the error here.

---

What is your concern?
This error is save to ignore because the lock has already been acquired and it doesn't matter whether the write operation to the s3 store fails, since the previous acquisition is still recorded.

---

Make sense.